src/main/kotlin/snyk/iac/annotator/IacBaseAnnotator.kt

```diff
@@ -0,0 +1,66 @@
+package snyk.iac.annotator
+
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.ExternalAnnotator
+import com.intellij.lang.annotation.HighlightSeverity
+import com.intellij.openapi.components.service
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.util.TextRange
+import com.intellij.openapi.util.io.FileUtil
+import com.intellij.psi.PsiFile
+import io.snyk.plugin.ui.toolwindow.SnykToolWindowPanel
+import snyk.iac.IacIssue
+import snyk.iac.IacResult
+
+private val LOG = logger<IacBaseAnnotator>()
+
+abstract class IacBaseAnnotator : ExternalAnnotator<PsiFile, List<IacIssue>>() {
+
+    abstract fun textRange(psiFile: PsiFile, iacIssue: IacIssue): TextRange
+
+    override fun apply(psiFile: PsiFile, annotations: List<IacIssue>?, holder: AnnotationHolder) {
+        LOG.debug("Call apply on ${psiFile.name}")
+        if (annotations == null || annotations.isEmpty()) return
+
+        LOG.info("Received ${annotations.size} IacIssue annotations for ${psiFile.virtualFile.name}")
+        annotations.forEach { iacIssue ->
+            LOG.debug("-> ${iacIssue.id}: ${iacIssue.title}: ${iacIssue.lineNumber}")
+
+            holder.newAnnotation(HighlightSeverity.WARNING, annotationMessage(iacIssue))
+                .range(textRange(psiFile, iacIssue))
+                .create()
+        }
+    }
+
+    override fun collectInformation(psiFile: PsiFile): PsiFile = psiFile
+
+    override fun doAnnotate(psiFile: PsiFile?): List<IacIssue> {
+        LOG.debug("Calling doAnnotate on ${psiFile?.name}")
+
+        val iacResult = psiFile?.project?.service<SnykToolWindowPanel>()?.currentIacResult ?: return emptyList()
+        ProgressManager.checkCanceled()
+
+        return getAnnotationForFile(psiFile, iacResult)
+    }
+
+    private fun getAnnotationForFile(psiFile: PsiFile, iacResult: IacResult): List<IacIssue> {
+        LOG.debug("Calling getAnnotationForFile for ${psiFile.virtualFile?.path}")
+
+        val iacIssuesForFile = iacResult.allCliIssues?.firstOrNull { iacIssuesForFile ->
+            val psiFilePath = FileUtil.toSystemIndependentName(psiFile.virtualFile?.canonicalPath ?: psiFile.name)
```

Comment:
use absolutPath and keep it nullable?

Reply:
by Absolut path I mean getPath() that's also system independent, but don't Resolves all symbolic links as canonicalPath()
I'm not sure if we want to do such resolve?
Keep it nullable, cause using `psiFile.name` in case it's null - is not logically correct imho.